### DISCUSSION TITLE
Vizualisation Options

### DISCUSSION THREAD

**Discussion by wiesehahn** [r-lidar/lasR#93] [Ideas]
Hey, 

(warning: I feel this might be a dumb question as I havent thought about it in detail)
I am wondering do you have any thoughts on implementing some visualisation option?

I think this might not be your focus, as lasR is designed to PROCESS large amounts of data and visualization is a different topic. Also visualisations for large datasets are probably not easily possible. E.g. `pipeline = plot_pointcloud()` on a folder with multiple files might not make sense, requires overviews or some other vizualisation strategy.

Although lasR is not meant for experimenting, I think data discovery is often part of the process to built a pipeline, and in this case it would be nice to have some effective visualisation tool integrated. 

Currently we would have to run a pipeline which writes the data to disk and then load it to R with lidR (pointcloud), terra (dsm, dtm) etc.  or open it with some other software to vizualize it. Would it be possible and make sense to integrate some tool or add some function to pass data to another tool to visualize (intermediate) data (VPC/pointcloud/raster/mesh/bbox/...) via lasR without parsing it to R? 

I havent yet thought about related issues with pipelines, processed data etc., was just wondering if you had any thoughts on this.

**Comment by Jean-Romain** [comment]
This is not a dumb question at all and is definitively the main limitation of `lasR`. You are 100% right.

This is why I'm myself more prone to use `lidR`. In `lidR` you can run each stage one by one, redo some stage with other parameters, visualize, redo again, check. This is not possible in `lasR` and definitively not achievable within R. Looks at `cloudcompare` or `computree`: you have a UI, you can check boxes to display or not some point cloud or features, go backward in the pipeline with buttons. In R we are running scripts and we do not have an UI.

The goal of `lasR` is to be memory efficient and fast. I reached the goal. This implied to get data in an underlying hidden C++ data structure. This has a lot of cons when working in R. You identified some from the user side, but from the developer side I'm even more limited by how R is working. And I'm not going to write my own software.

Ideally my view is to design a processing pipeline with `lidR` and then be able to redo the exact same pipeline with `lasR`. `lidR` being the R&D and `lasR` being the production. I can do it because I have the ability to change the code the way I want. But remember than both `lidR` and `lasR` were primarily designed for my own use. And yes `lidR` and `lasR`are meant to be used together .

Your suggestion `pipeline = plot_pointcloud()` is not dumb at all and I'm considering it actually. Something like:

```r
exec(pipeline, on = f, viewer = TRUE)
```

That would open a viewer window that display in live what append in each stage. With the possibility to stop the pipeline between each stage to get the time to visualize. A lot of work for an unknown added value.

**Comment by wiesehahn** [comment]
Puh, I am delighted its not a dumb question, so I can try with more possibly dumb comments. 

Not having an UI with buttons might be a drawback but I like the advantage it brings in terms of reproducibility. 

Thanks for your explanation, it is a little bit clearer now what is intended. lidRs plotting functionality works in general but is also limited due to rgl as far as I understand. I am wondering if it would be possible to pass data in your structure to existing software to vizualize it more effectively. 

I am technically not that much into it, but have you thought about Apache Arrow/Geoarrow as in memory data structure? It seems to be a thing nowadays to work with e.g. arrow and duckDB. I think in theory it also allows to work with data across languages. PDAL has some read/write capabilities for it (https://pdal.io/en/latest/stages/writers.arrow.html). Regarding point cloud vizulalisation this might enable effective vizualisation in the browser via deckgl (simialar to what [lonboard](https://developmentseed.org/blog/2023-10-23-lonboard/) does for vector data). 

But of course there are so many rendering tools its impossible to keep track (https://github.com/potree/potree, https://github.com/m-schuetz/compute_rasterizer, https://github.com/ComputingElevatedLab/WGPU-COPC-Viewer, https://github.com/iTowns/itowns, ...)

**Comment by Jean-Romain** [comment]
> Not having an UI with buttons might be a drawback but I like the advantage it brings in terms of reproducibility.

100% agree

>  lidRs plotting functionality works in general but is also limited due to rgl as far as I understand.

https://github.com/r-lidar-lab/lidRviewer/issues/23

> Apache Arrow/Geoarrow as in memory data structure? 

I don't know about Arrow actually. What I know however is that we are limited in R. I could do almost anything on my side on my own computer. I could use `Arrow`, `PCL`, `CGAL` in an R package  *on my own computer* but then to make it available as an easy to install package working on Linux, Mac and Windows I'm limited by the CRAN policy and what is available in Rtools. So it is a trade-off between all that R brings (e.g. `sf`, `terra` that allow me to load and visualize raster and vector effortlessly) and the limitation to what kind of library I can use du to CRAN policies. I could integrate `PCL` based tools but then 99.9% of people would no longer be able to install the packages.
